node/coinstacks/solana/api/src/websocket.ts

```diff
@@ -0,0 +1,107 @@
+import { Logger } from '@shapeshiftoss/logger'
+import { WebsocketClient as BaseWebsocketClient, Args, Options, Subscription } from '@shapeshiftoss/websocket'
+import WebSocket from 'ws'
+import { Transaction, WebsocketResponse } from './types'
+
+const logger = new Logger({ namespace: ['unchained', 'solana', 'websocket'], level: process.env.LOG_LEVEL })
+
+interface WebsocketArgs extends Omit<Args, 'logger'> {
+  transactionHandler: TransactionHandler | Array<TransactionHandler>
+}
+
+type TransactionHandler = (data: Transaction) => Promise<void>
+
+export class WebsocketClient extends BaseWebsocketClient {
+  private handleTransaction: TransactionHandler | Array<TransactionHandler>
+
+  private addresses: Array<string> = []
+  private subscriptionsIds: Array<number> = []
+
+  constructor(_url: string, args: WebsocketArgs, opts?: Options) {
+    const url = args.apiKey ? `${_url}?api-key=${args.apiKey}` : _url
+    super(url, { logger }, opts)
+
+    this.handleTransaction = args.transactionHandler
+
+    this.initialize()
+
+    this.socket.onmessage = (msg) => this.onMessage(msg)
+    this.socket.onopen = () => this.onOpen()
+  }
+
+  protected onOpen(): void {
+    super._onOpen()
+
+    if (this.addresses.length > 0) {
+      const subscribeAddresses = this.getAddressesSubscription()
+      this.socket.send(JSON.stringify(subscribeAddresses))
+    }
+  }
+
+  protected async onMessage(message: WebSocket.MessageEvent): Promise<void> {
+    try {
+      const res: WebsocketResponse = JSON.parse(message.data.toString())
+
+      if (res.params?.subscription && !this.subscriptionsIds.includes(res.params.subscription)) {
+        this.subscriptionsIds = [...this.subscriptionsIds, res.params.subscription]
+      }
+
+      switch (res.method) {
+        case 'transactionNotification':
+          if ('result' in res.params) {
+            if (Array.isArray(this.handleTransaction)) {
+              this.handleTransaction.map(async (handleTransaction) => handleTransaction(res.params.result.transaction))
+            } else {
+              this.handleTransaction(res.params.result.transaction)
+            }
+          }
+          return
+      }
+    } catch (err) {
+      this.logger.error(err, `failed to handle message: ${JSON.stringify(message)}`)
+    }
+  }
+
+  subscribeAddresses(addresses: string[]): void {
+    this.addresses = addresses
+    const subscribeAddresses = this.getAddressesSubscription()
+
+    this.subscriptionsIds.forEach((id) => {
```

Comment:
my only thought is due to the async nature of the websockets, we may have back to back `subscribeAddress` calls before we handle any `onMessage` results leading to multiple active subscriptions possibly. A thought would be to have an incrementor for our subscription ids so that we can handle the `onMessage` subscription result and unsubscribe/ignore any subscriptions that are < currentSubscriptionId.

Reply:
I'll have to think a bit more about this one, it makes a lot of sense, with a wide usage, it can be an issue for sure

Reply:
Ok cool. Historically websockets are the flakiest part of our app, so would be good to ensure we are handling weird edge cases as best as possible. Let me know if you want to brainstorm at all on it!

Reply:
Should be handled in [6226994](https://github.com/shapeshift/unchained/pull/1053/commits/62269940cd6fb501b964597b35e4dc13117ed672)